### DISCUSSION TITLE
Export Contacts to vCards

### DISCUSSION THREAD

**Discussion by loregood** [netbox-community/netbox#10586] [Ideas]
I'm looking at creating an Export Template of Contact information to a vCard file for easy import into many different cross-platform Contact applications.

I'm no expert in vCards, but I have looked at the "vCard Format Specification" described in [RFC-6350](https://www.rfc-editor.org/rfc/rfc6350.html).

I think that 2 changes could improve export (and import) to and from vCards significantly.

1. A `"uuid":` field for `tenancy/contacts` in the URN namespace defined in [RFC4122](https://www.rfc-editor.org/rfc/rfc4122) created (with a unique UID) upon creation. Used for the `UID` field described in [section-6.7.6](https://www.rfc-editor.org/rfc/rfc6350.html#section-6.7.6) to uniquely identify a vCard.
```
    "uuid": "03a0e51f-d1aa-4385-8a53-e29025acd8af",
``` 
2. A more structured output for the `"address":` field (like a JSON object) for all the components describing an address. Used for the `ADR` field described in [section-6.3.1](https://www.rfc-editor.org/rfc/rfc6350.html#section-6.3.1). The values can be empty.
```
    "address": {
        "po_box": "the post office box",
        "ext_addr": "the extended address (e.g., apartment or suite number)",
        "street": "the street address",
        "city": "the locality (e.g., city)",
        "region": "the region (e.g., state or province)",
        "postal": "the postal code",
        "country": "the country name"
    },
```

A simple Export Template of all Contacts (`Content Type: tenancy | contact`, 
`MIME Type: text/vcard`, `File Extension: vcf` as an Attachment) tested on NetBox v3.2.9:
```
{# RFC-6350 vCard export of all Contacts to a single .vcf file -#}
{# Export to VERSION:3.0 (better supported when importing) -#}
{%  for contact in queryset %}
BEGIN:VCARD
VERSION:3.0
KIND:individual
N:;;;;
FN;CHARSET=UTF-8:{{ contact.name }}
{%      if contact.title -%}
TITLE;CHARSET=UTF-8:{{ contact.title }}
{%      endif -%}
{%      if contact.phone -%}
TEL;type=WORK;type=VOICE:{{ contact.phone }}
{%      endif -%}
{%      if contact.cf.mobile -%}
TEL;type=CELL;type=VOICE:{{ contact.cf.mobile }}
{%      endif -%}
{%      if contact.email -%}
EMAIL;type=INTERNET;type=WORK;CHARSET=UTF-8:{{ contact.email }}
{%      endif -%}
{%      if contact.address -%}
{%          set ESC_COMMA = contact['address'] | replace(",", "\,") -%}
{%          set ESC_DASH = ESC_COMMA | replace("-", "\-") -%}
{%          set ESC_NEWLINE = ESC_DASH | replace("\r\n", ";") -%}
ADR;type=WORK;CHARSET=UTF-8:;;{{ ESC_NEWLINE }}
{%      endif -%}
END:VCARD
{%  endfor %}
```

With the `"address":` field currently formatet like this:
```
"address": "First Line\r\nSecond Line\r\nMaybe more lines..?",
```
It's really hard (guessing required) to create a parser in Jinja that correctly handles how a user might type in an address. It doesn't get easier when you have addresses in multiple countries (with different formats). The `ADR` must have them is the format below (all the fields separated with `;`) and in that order:
```
{# ADR RFC: "PO_BOX;EXT_ADDR;STREET;CITY;REGION;POSTAL;COUNTRY" -#}
```
Some countries don't use `REGION` and `POSTAL` might be written before `CITY`. If a field is empty you still need the `;` separator. So a line might look something like: `;;STREET;CITY;;POSTAL;`

 For better support with multiple different Contact applications you generally want to keep `PO_BOX` and `EXT_ADDR` empty - so your ADR line should probably start with `ADR:;;`

I'm not sure how much work the changes above would require and wondered if anybody else would see any value in better support for vCard import and export of Contacts to and from NetBox?

**Comment by candlerb** [comment]
As a first step, or a prototype/proof-of-concept, you could add Custom Fields to Tenancy > Contact, containing the uuid and the structured address fields.
